pybossa/sched.py

```diff
@@ -267,16 +270,32 @@ def template_get_locked_task(project_id, user_id=None, user_ip=None,
         sql = query_factory(project_id, user_id=user_id, limit=limit,
                             rand_within_priority=rand_within_priority,
                             task_type=task_type)
+        db_max_records = 100000
+        limit = db_max_records if filter_user_prefs else user_count + 5
         rows = session.execute(sql, dict(project_id=project_id,
                                          user_id=user_id,
                                          assign_user=assign_user,
-                                         limit=user_count + 5))
+                                         limit=limit))
+        user_profile = cached_users.get_user_profile_metadata(user_id)
+        if filter_user_prefs:
+            # validate user qualification and calculate task preference score
+            user_profile = json.loads(user_profile) if user_profile else {}
+            task_rank_info = []
+            for task_id, taskcount, n_answers, calibration, w_filter, w_pref, timeout in rows:
+                w_pref = w_pref or {}
+                w_filter = w_filter or {}
+                meet_requirement = cached_helpers.user_meet_task_requirement(task_id, w_filter, user_profile)
+                if meet_requirement:
+                    score = get_task_preference(w_pref, user_profile)
+                    task_rank_info.append((task_id, taskcount, n_answers, calibration, score, None, timeout))
+            rows = sorted(task_rank_info, key=lambda tup: tup[4], reverse=True)
+        else:
+            rows = [r for r in rows]
 
-        for task_id, taskcount, n_answers, calibration, timeout in rows:
+        for task_id, taskcount, n_answers, calibration, _, _, timeout in rows:
             timeout = timeout or TIMEOUT
             remaining = float('inf') if calibration else n_answers - taskcount
             if acquire_lock(task_id, user_id, remaining, timeout):
```

Comment:
After obtaining task, the db cusor wasn't closed as it was originally. Is this intentional and would cover both the cases with and without filter_user_pref?

Reply:
This for loop doesn't use db cursor anymore as `rows` is a list

Reply:
Its because you're using the same variable [here](https://github.com/bloomberg/pybossa/pull/595/files#diff-f7eb1704644fb813492849abfa5e3f07f4c3cd76c4f60088485e5ffe204b5024R293). I believe if db cursor is not closed, it would be db resource leak. Hence something on similar lines could be done
```python
records = [r for r in rows]
rows.close()
for task_id, taskcount, n_answers, calibration, _, _, timeout in records:
    ...

Reply:
We don't need to explicitly close the cursor as it is closed when the for loop ends. See https://docs.sqlalchemy.org/en/13/core/connections.html/
"The DBAPI cursor will be closed by the ResultProxy when all of its result rows (if any) are exhausted. A ResultProxy that returns no rows, such as that of an UPDATE statement (without any returned rows), releases cursor resources immediately upon construction."

Reply:
Thanks for confirming. seems old rows.close() was unnecessary?

Reply:
Old rows.close() is needed as it stops (return) before exhausting all data in for loop